src/compiler/checker.ts

```diff
@@ -16052,9 +16176,13 @@ namespace ts {
                     if (produceDiagnostics) {
                         const symbol = getNodeLinks(node).resolvedSymbol;
                         const typeParameters = symbol.flags & SymbolFlags.TypeAlias ? getSymbolLinks(symbol).typeParameters : (<TypeReference>type).target.localTypeParameters;
-                        checkTypeArgumentConstraints(typeParameters, node.typeArguments);
+                        const minTypeArgumentCount = symbol.flags & SymbolFlags.TypeAlias ? getSymbolLinks(symbol).minTypeArgumentCount : (<TypeReference>type).target.minTypeArgumentCount;
+                        checkTypeArgumentConstraints(typeParameters, node.typeArguments, minTypeArgumentCount);
                     }
                 }
+                if (type.flags & TypeFlags.TypeParameter && !(<TypeParameter>type).isThisType && type.symbol && !isTypeParameterInScope(<TypeParameter>type, node)) {

```

Comment:
That is what this is trying to accomplish. `isTypeParameterInScope` walks up the spine from `node` to find a declaration that contains the type parameter. Are you stating that this condition is too permissive or that there is a better approach?